src/compartment.c

```diff
@@ -436,28 +544,48 @@ comp_map(struct Compartment* to_map)
     for (size_t i = 0; i < to_map->seg_count; ++i)
     {
         curr_seg = to_map->segs[i];
-        if (curr_seg->file_sz == curr_seg->mem_sz)
+        map_result = mmap((void*) curr_seg->mem_bot,
+                                curr_seg->mem_sz,
+                                /*curr_seg->prot_flags,*/ // TODO currently need read/write to inject the intercepts, consider better option
+                                PROT_READ | PROT_WRITE | PROT_EXEC,
+                                MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+                                -1, 0);
+        if (map_result == MAP_FAILED)
         {
-            map_result = mmap((void*) curr_seg->mem_bot,
-                                    curr_seg->file_sz,
-                                    /*curr_seg->prot_flags,*/ // TODO currently need read/write to inject the intercepts, consider better option
-                                    PROT_READ | PROT_WRITE | PROT_EXEC,
-                                    MAP_PRIVATE | MAP_FIXED,
-                                    to_map->fd, curr_seg->offset);
+            perror("Error mapping comp segment");
+            exit(1);
         }
-        else
+        printf("HERE COMP\n");
+        do_pread(to_map->fd, (void*) curr_seg->mem_bot, curr_seg->file_sz,
+                 curr_seg->offset);
+    }
+
+    // Map compartment library dependencies segments
+    struct LibDependency* lib_dep;
+    struct SegmentMap* lib_dep_seg;
+    int lib_dep_fd;
+    for (size_t i = 0; i < to_map->lib_deps_count; ++i)
+    {
+        lib_dep = to_map->lib_deps[i];
+        lib_dep_fd = open(lib_dep->lib_path, O_RDONLY);
+        for (size_t j = 0; j < lib_dep->lib_segs_count; ++j)
         {
-            assert(curr_seg->mem_sz > curr_seg->file_sz);
-            map_result = mmap((void*) curr_seg->mem_bot,
-                                    curr_seg->mem_sz,
-                                    curr_seg->prot_flags,
-                                    MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
-                                    -1, 0);
-            assert(map_result !=  MAP_FAILED);
-            int pread_res = pread(to_map->fd, (void*) curr_seg->mem_bot,
-                                  curr_seg->file_sz, curr_seg->offset);
-            assert(pread_res != -1);
+            lib_dep_seg = lib_dep->lib_segs[j];
+            map_result = mmap((void*) lib_dep->lib_mem_base + lib_dep_seg->mem_bot,
+                              lib_dep_seg->mem_sz,
+                              PROT_READ | PROT_WRITE | PROT_EXEC, // TODO fix
+                              MAP_PRIVATE | MAP_FIXED | MAP_ANONYMOUS,
+                              -1, 0);
+            if (map_result == MAP_FAILED)
+            {
+                perror("Error mapping library dependency segment");
```

Comment:
Sometimes we use `printf ; exit` sometimes `perror ; exit`. Can we just use `err` or similar?

Reply:
All of these should be now fixed as discussed earlier. This one should've been done in 569506d85d7900debb57337fdd785962cd4d02d2.